pallets/crowdloan-rewards/src/lib.rs

```diff
@@ -0,0 +1,335 @@
+		pub last_paid: T::BlockNumber,
+	}
+
+	// No hooks
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {}
+
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Associate a native rewards_destination identity with a crowdloan contribution.
+		///
+		/// This is an unsigned call because the caller may not have any funds to pay fees with.
+		/// This is inspired by Polkadot's claims pallet:
+		/// https://github.com/paritytech/polkadot/blob/master/runtime/common/src/claims.rs
+		///
+		/// This function and the entire concept of unassociated contributions may be obviated if
+		/// They will accept a memo filed in the Polkadot crowdloan pallet.
+		#[pallet::weight(0)]
+		pub fn associate_native_identity(
+			origin: OriginFor<T>,
+			reward_account: T::AccountId,
+			relay_account: T::RelayChainAccountId,
+			proof: MultiSignature,
+		) -> DispatchResultWithPostInfo {
+			ensure_none(origin)?;
+			//TODO check the proof:
+			// 1. Is signed by an actual unassociated contributor
+			// 2. Signs a valid native identity
+			// Check the proof. The Proof consists of a Signature of the rewarded account with the
+			// claimer key
+			let payload = reward_account.encode();
+			ensure!(
+				proof.verify(payload.as_slice(), &relay_account.clone().into()),
+				Error::<T>::InvalidClaimSignature
+			);
+
+			// We ensure the mapping does not exist yet to avoid multi-claiming
+			ensure!(
+				AccountsMapping::<T>::get(&relay_account).is_none(),
+				Error::<T>::AlreadyAssociated
+			);
+
+			// Upon error this should check the relay chain state in this case
+			let reward_info = UnassociatedContributions::<T>::get(&relay_account)
+				.ok_or(Error::<T>::NoAssociatedClaim)?;
+
+			// Insert on payable
+			AccountsPayable::<T>::insert(&reward_account, &reward_info);
+
+			// Remove from unassociated
+			<UnassociatedContributions<T>>::remove(&relay_account);
+
+			// Insert in mapping
+			AccountsMapping::<T>::insert(&relay_account, &reward_account);
+
+			// Emit Event
+			Self::deposit_event(Event::NativeIdentityAssociated(
+				relay_account,
+				reward_account,
+				reward_info.total_reward,
+			));
+
+			Ok(Default::default())
+		}
+
+		/// Collect whatever portion of your reward are currently vested.
+		#[pallet::weight(0)]
+		pub fn show_me_the_money(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
+			let payee = ensure_signed(origin)?;
+
+			// Calculate the veted amount on demand.
+			let mut info =
+				AccountsPayable::<T>::get(&payee).ok_or(Error::<T>::NoAssociatedClaim)?;
+			let now = frame_system::Module::<T>::block_number();
+			//TODO This part doesn't compile because of a million stupid errors about converting
+			// between u32, Balance, and BlockNumber. I think that is solvable, just annoying.
+			let payable_per_block = info.total_reward
+				/ T::VestingPeriod::get()
+					.saturated_into::<u128>()
+					.try_into()
+					.ok()
+					.ok_or(Error::<T>::WrongConversionU128ToBalance)?; //TODO safe math;
```

Comment:
It seems like this is ripe for rounding errors (although maybe a moot point with high decimal precision).

A very contrived example might be:
* reward = 4
* blocks = 3
* payable_per_block would truncate to 1
* accounted for equally over each block, 3 < 4

If, instead, we stored the total reward and the total paid out, the logic becomes:

* calculate total rewardable: `(vested_blocks / vesting_length) * total_reward`
* calculate `currently_rewardable = total_reward - total_paid`
* pay out currently_rewardable and update `total_paid += currently_rewardable`